### ISSUE TITLE
Can't connect using X509

### ISSUE THREAD

**Issue by cadamis** [elixir-mongo/mongodb#239]
I've got an AWS production DB I want to connect to, but I get `[error] Mongo.Protocol (#PID<0.208.0>) failed to connect: ** (Mongo.Error) ssl connect: TLS connection is closed - :closed` even when using the exact params that work from my python script. The general idea is this:
```
username = "CN=my.username,OU=Clients,O=MyOrg,L=MyCity,ST=MyState,C=US"
{:ok, conn} = Mongo.start_link(
      database: "mydatabase",
      hostname: "db.mycompany.com",
      username: username,
      ssl: true,
      auth_mechanism: "MONGODB-X509",
      ssl_opts: [
        cacertfile: "/home/myuser/.mongo/root-ca.pem",
        certfile: "/home/myuser/.mongo/my-key.pem"
      ]
    )
```

I've tried this with and without `versions: [:"tlsv1.2"]` in the ssl_opts, doesn't seem to make a difference.

I first tried to connect using the `url` param with a hostname like this:
`mongodb://{#username}@db.mycompany.com/mydatabase?authMechanism=MONGODB-X509`
I noticed there was a problem with the `@mongo_url_regex` regex in `mongodb/lib/mongo/url_parser.ex` that ended up mangling the host, I assume because of the LDAP-formatted username. After specifying everything in params instead, as shown above, it still won't connect, and I don't know where else to look for more specific errors.

I know you tracked down a problem with Erlang SSL once already, I'm wondering if this goes all the way back to that lib. I'd love to help track it down if you have any ideas on where else to look for errors about the connection.

**Comment by ankhers** [comment]
I have not looked into getting the x509 authentication mechanism working before. Officially this driver supports CR (which is deprecated and actually removed in 4.0) and SCRAM. I can try looking into this when I find a bit of time.

**Comment by cadamis** [comment]
Thanks for the quick response. I'm new to Elixir, but I think the RabbitMQ stuff has X509 working. I'll see if I can use that as an example and poke at it.

**Comment by cadamis** [comment]
I was able to dig around and get something working. I think it's fairly consistent with the rest of the code (hopefully). Would you like me to put together a PR?

**Comment by ankhers** [comment]
Yes please.

**Comment by cadamis** [comment]
PR: #242

**Comment by ankhers** [comment]
Merged in #242
